pkg/lightning/backend/external/onefile_writer.go

```diff
@@ -203,5 +308,21 @@ func (w *OneFileWriter) closeImpl(ctx context.Context) (err error) {
 		err = err2
 		return
 	}
+	if w.dupWriter != nil {
+		w.dupKVStore.finish()
+		if err3 := w.dupWriter.Close(ctx); err3 != nil {
+			w.logger.Error("Close dup writer failed", zap.Error(err3))
+			err = err3
+			return
+		}
+	}
 	return nil
 }
+
+func encodeToBuf(buf, key, value []byte) {
+	keyLen := len(key)
+	binary.BigEndian.AppendUint64(buf[:0], uint64(keyLen))
+	binary.BigEndian.AppendUint64(buf[lengthBytes:lengthBytes], uint64(len(value)))
+	copy(buf[lengthBytes*2:], key)
+	copy(buf[lengthBytes*2+keyLen:], value)
+}
```

Comment:
can we pre allocate the buf?
```suggestion
    keyLen := len(key)
    valueLen := len(value)
 	data := make([]byte, 0, lengthBytes*2+keyLen+valueLen)
 	data = binary.BigEndian.AppendUint64(data, uint64(keyLen))
 	data = binary.BigEndian.AppendUint64(data, uint64(valueLen))
 	data = append(data, key...)
 	data = append(data, value...)
 	copy(buf, data)
```